pageserver/src/tenant/timeline.rs

```diff
@@ -3015,6 +3017,37 @@ impl Timeline {
             .read_lock_held_spawn_blocking_startup_micros
             .till_now();
 
+        // before we start, because we can produce duplicate L1 layers, we must first collect all
+        // existing L1's overlapping with the LSN range, and make sure they are resident for the
+        // duration of the compaction. any duplicate Layers will end up touching the same
+        // filesystem file on eviction, and we do not want that to happen.
+        let resident_overlapping_l1s = {
```

Comment:
Merging in contents from comment inside this block.
And, flip the order, describing problem first, then solution.


```suggestion
        // there is a bug where compaction will overwrite an existing an L1
        // (see https://github.com/neondatabase/neon/pull/5086).
        // if the existing L1 is evicted, the following race condition can happen:
        // 1. existing L1 layer is currently evicted
        // 2. compaction: due to above bug, re-compacts a bunch of L0s, producing a layer file with
              the same key & lsn range as the existing L1, with same logical contents, but possibly
              physically different bit pattenr / file size.
        // 3. comapction: gets preempted before it creates the corresponding Layer object
        // 4. another thread: on-demand downloads the existing L1 from S3, thereby overwriting the
              file the that compaction just produced
        // 5. compaction: resumes, creates Layer object based on what it thinks it wrote out to disk,
        //    whereas in reality, the on-demand download has overwritten the file
        //
        // The workaround here is to make any existing L1s that this compaction may overwrite resident
        // for the duration of the compaction.
        let resident_overlapping_l1s = {
```